File: Problems/Algorithms/1007. Minimum Domino Rotations for Equal Row/MinimumRotations.kt
class Solution {
    fun minDominoRotations(tops: IntArray, bottoms: IntArray): Int {
        val map = mutableMapOf<Int, Int>()
        val n = tops.size
        
        for (i in 0..n-1) {
            map.put(tops[i], map.getOrDefault(tops[i], 0) + 1)
            map.put(bottoms[i], map.getOrDefault(bottoms[i], 0) + 1)
        }
        
        var max = 0
        var num = 0
        for (key in map.keys) {
            if (map[key]!! > max) {
                max = map[key]!!
                num = key
            }
        }
        
        if (max < n) return -1
        
        var topn = 0
        var bottomn = 0
        for (i in 0..n-1) {
            if (tops[i] != num && bottoms[i] != num) return -1
            
            if (tops[i] != num) {
                topn++
            } else if (bottoms[i] != num) {
                bottomn++
            }
        }
        
        return minOf(topn, bottomn)
    }
}

File: Problems/Algorithms/1007. Minimum Domino Rotations for Equal Row/minimum_rotations.py
#!/usr/bin/env python3

from typing import List


class Solution:
    def minDominoRotations(self, tops: List[int], bottoms: List[int]) -> int:
        ans = self.rotate(tops, bottoms, tops[0])
        if tops[0] != bottoms[0]:
            ans = min(ans, self.rotate(tops, bottoms, bottoms[0]))

        return -1 if ans == float('inf') else ans

    def rotate(self, tops, bottoms, target):
        rotate_top = rotate_bottom = 0

        for i in range(len(tops)):
            if tops[i] != target and bottoms[i] != target:
                return float('inf')
            if tops[i] != target:
                rotate_top += 1
            if bottoms[i] != target:
                rotate_bottom += 1

        return min(rotate_top, rotate_bottom)
